control/setup_control.go

```diff
@@ -0,0 +1,398 @@
+	NeedsAutoTuning bool
+
+	// LoopFrequency is the frequency at which the control loop should run
+	LoopFrequency float64
+
+	// ControllableType is the type of component the control loop will be set up for,
+	// currently a base or motor
+	ControllableType string
+
+	// NeedsSingleAutoTuning is true when, in the case of multiple PID blocks, only one of
+	// them needs to be auto-tuned
+	NeedsSingleAutoTuning bool
+}
+
+// SetupPIDControlConfig creates a control config.
+func SetupPIDControlConfig(
+	pidVals []PIDConfig,
+	componentName string,
+	options Options,
+	c Controllable,
+	logger logging.Logger,
+) (*PIDLoop, error) {
+	pidLoop := &PIDLoop{
+		Controllable: c,
+		logger:       logger,
+		Options:      options,
+		ControlConf:  Config{},
+		ControlLoop:  nil,
+	}
+
+	// set controlConf as either an optional custom config, or as the default control config
+	if options.UseCustomConfig {
+		pidLoop.ControlConf = options.CompleteCustomConfig
+		for i, b := range options.CompleteCustomConfig.Blocks {
+			if b.Type == blockSum {
+				sumIndex = i
+			}
+		}
+	} else {
+		pidLoop.createControlLoopConfig(pidVals, componentName)
+	}
+
+	// auto tune the control loop if needed
+	if options.NeedsAutoTuning {
+		cancelCtx, cancelFunc := context.WithCancel(context.Background())
+		if err := pidLoop.TunePIDLoop(cancelCtx, cancelFunc); err != nil {
+			return nil, err
+		}
+	}
+
+	if options.NeedsSingleAutoTuning {
+		if err := pidLoop.StartControlLoop(); err != nil {
+			return nil, err
+		}
+	}
+
+	return pidLoop, nil
+}
+
+// TunePIDLoop runs the auto-tuning process for a PID control loop.
+func (p *PIDLoop) TunePIDLoop(ctx context.Context, cancelFunc context.CancelFunc) error {
+	var errs error
+	p.activeBackgroundWorkers.Add(1)
+	utils.PanicCapturingGo(func() {
+		defer func() {
+			cancelFunc()
+			if p.ControlLoop != nil {
+				p.ControlLoop.Stop()
+				p.ControlLoop = nil
+			}
+		}()
+		defer p.activeBackgroundWorkers.Done()
+		select {
+		case <-ctx.Done():
+			return
+		default:
+		}
+		// switch sum to depend on the setpoint if position control
+		if p.Options.PositionControlUsingTrapz {
+			p.ControlConf.Blocks[sumIndex].DependsOn[0] = p.BlockNames["constant"][0]
+			if err := p.StartControlLoop(); err != nil {
+				errs = multierr.Combine(errs, err)
+			}
+
+			if err := p.ControlLoop.MonitorTuning(ctx); err != nil {
+				errs = multierr.Combine(errs, err)
+			}
+		}
+		if p.Options.SensorFeedback2DVelocityControl {
+			// to tune linear PID values, angular PI values must be non-zero
+			p.ControlConf.Blocks[angularPIDIndex].Attribute["kP"] = 0.0001
+			p.ControlConf.Blocks[angularPIDIndex].Attribute["kI"] = 0.0001
+			p.logger.Info("tuning linear PID")
+			if err := p.StartControlLoop(); err != nil {
+				errs = multierr.Combine(errs, err)
+			}
+
+			if err := p.ControlLoop.MonitorTuning(ctx); err != nil {
+				errs = multierr.Combine(errs, err)
+			}
+
+			p.ControlLoop.Stop()
+			p.ControlLoop = nil
```

Comment:
possible refactor to remove `NeedsSingleAutoTuning`. Still kinda gross but I think this does what you wanted, and moves all tuning logic into one parameter

```suggestion
 // check if linear needs to be tuned
 if p.ControlConf.Blocks[linearPIDIndex].Attribute["kP"] == 0.0 && p.ControlConf.Blocks[linearPIDIndex].Attribute["kPI"] = 0.0{
 // preserve angular values and set them to be non-zero
 kPA := p.ControlConf.Blocks[angularPIDIndex].Attribute["kP"]
 kIA := p.ControlConf.Blocks[angularPIDIndex].Attribute["kI"]
			// to tune linear PID values, angular PI values must be non-zero
			p.ControlConf.Blocks[angularPIDIndex].Attribute["kP"] = 0.0001
			p.ControlConf.Blocks[angularPIDIndex].Attribute["kI"] = 0.0001
			p.logger.Info("tuning linear PID")
			if err := p.StartControlLoop(); err != nil {
				errs = multierr.Combine(errs, err)
			}

			if err := p.ControlLoop.MonitorTuning(ctx); err != nil {
				errs = multierr.Combine(errs, err)
			}

			p.ControlLoop.Stop()
			p.ControlLoop = nil
			
			// reset angular values
			p.ControlConf.Blocks[angularPIDIndex].Attribute["kP"] = kPA
 p.ControlConf.Blocks[angularPIDIndex].Attribute["kI"] = kIA
}

// do the same process for linear
```